core/src/main/scala/io/qbeast/spark/index/SparkPlanAnalyzer.scala

```diff
@@ -0,0 +1,109 @@
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.plans.logical.Project
+import org.apache.spark.sql.catalyst.plans.logical.Sample
+import org.apache.spark.sql.catalyst.plans.logical.Sort
+import org.apache.spark.sql.DataFrame
+
+import scala.collection.convert.ImplicitConversions.`collection asJava`
+
+object SparkPlanAnalyzer {
+
+  /**
+   * Builds a list of non-deterministic operations in the logical plan
+   *
+   * A list of Non-Deterministic operations:
+   *   - LocalLimit without Sort Operation
+   *   - Sample
+   *   - Filter with non-deterministic condition
+   *
+   * @param plan
+   *   the logical plan
+   * @return
+   */
+  private def isLogicalPlanDeterministic(plan: LogicalPlan): Boolean = {
+    // Recursively traverse the logical plan to find non-deterministic operations
+    val isCurrentOperationDeterministic = plan match {
+      case LocalLimit(_, _: Sort) => true // LocalLimit with Sort is deterministic
+      case LocalLimit(_, _) => false
+      case Sample(_, _, _, _, _) => false
+      case Filter(condition, _) => condition.deterministic
+      case _ => true
+    }
+
+    val areChildOperationsDeterministic = plan.children.forall(isLogicalPlanDeterministic)
+
+    isCurrentOperationDeterministic && areChildOperationsDeterministic
+  }
+
+  /**
+   * Extracts the select expressions from the logical plan for the provided column name
+   *
+   * @param logicalPlan
+   *   the logical plan
+   * @param columnName
+   *   the column name
+   * @return
+   */
+  private[index] def collectSelectExpressions(
+      logicalPlan: LogicalPlan,
+      columnName: String): Seq[Expression] = {
+    logicalPlan match {
+      case Project(projectList, _) =>
+        projectList.collect {
+          case Alias(child, name) if name == columnName => child
+          case expr if expr.references.map(_.name).contains(columnName) => expr
+        }
+      case _ =>
+        logicalPlan.children.flatMap(child => collectSelectExpressions(child, columnName))
+    }
+  }
+
+  /**
+   * Checks if a column is deterministic in the logical plan
+   * @param logicalPlan
+   *   the logical plan
+   * @param columnName
+   *   the column name
+   * @return
+   */
+
+  private def isColumnDeterministic(logicalPlan: LogicalPlan, columnName: String): Boolean = {
+    val expressionSet = collectSelectExpressions(logicalPlan, columnName)
+    expressionSet.forall(_.deterministic)
+  }
+
+  /**
+   * Analyzes the DataFrame to determine if it's execution is safely deterministic for indexing
+   *
+   *   - The logical plan of the DataFrame is checked for determinism
+   *   - The columns to index are checked for determinism
+   *
+   * @param dataFrame
+   *   the DataFrame to analyze
+   * @param columnsToIndex
+   *   the columns to index
+   * @return
+   */
+  def analyzeDataFrameDeterminism(dataFrame: DataFrame, columnsToIndex: Seq[String]): Boolean = {
```

Comment:
I want to propose the following:
```
def assertDataFrameDeterminism(dataFrame: DataFrame, revision: Revision): Boolean
```
This should be the first thing we do when entering `DataAnalyzer.analyze,` it should throw an error inside the method if the conditions aren't met.

Reply:
Two things here:

1. the api makes sense, in that way we can have a broader set of Revision characteristics to tests.
2. However I am wondering about the second statement...
> This should be the first thing we do when entering DataAnalyzer.analyze. 

What if the user changes the columns to index in the future? Wouldn't we need to asses that first with the `SparkRevisionFactory`?

Reply:
We can reverse the order if column changes ever come to be, but for now, I think it'd be smart to avoid computing the dataframe statistics if the process is to be aborted.